src/vm/eventpipe.cpp

```diff
@@ -125,37 +126,33 @@ void EventPipe::Shutdown()
         return;
     }
 
+    if (s_state != EventPipeState::Initialized)
+        return;
+
     // We are shutting down, so if disabling EventPipe throws, we need to move along anyway.
     EX_TRY
     {
-        bool tracingInitialized = false;
         {
             CrstHolder _crst(GetLock());
-            tracingInitialized = s_tracingInitialized;
-
-            // Mark tracing as no longer initialized.
-            s_tracingInitialized = false;
+            s_state = EventPipeState::ShuttingDown;
```

Comment:
There are enough state values possible for `EventPipeState` that it might be useful to make the state transitions a bit more rigid through a helper that sets `s_state` to the desired state and performs checks to make sure the transition is valid. Ie, transitioning from `Initialized -> ShuttingDown` is legal, but `ShuttingDown -> Initialized` would be invalid.

Reply:
With the simpler NotInitialized -> Initialized -> ShuttingDown progression I'm fine with or without a helper, whatever you guys decide.